### ISSUE TITLE
[Question]: How can we use exceptions that have input arguments in constructor with Templated Exceptions

### ISSUE THREAD

**Issue by kasvith** [nanogiants/nestjs-swagger-api-exception-decorator#33]
Thank you for super useful package

I've defined my `TemplatedException` as follows. 

```ts
// templated-api-exception.ts

import {
  buildPlaceholder,
  buildTemplatedApiExceptionDecorator,
} from '@nanogiants/nestjs-swagger-api-exception-decorator';
import { BaseException } from 'common/exceptions';

export const TemplatedApiException = buildTemplatedApiExceptionDecorator(
  {
    message: '$message',
    statusCode: '$status',
    errorCode: '$errorCode',
    error: '$error',
    errors: '$errors',
  },
  {
    requiredProperties: ['statusCode', 'message', 'error', 'errorCode'],
    placeholders: {
      errors: buildPlaceholder(
        () => BaseException,
        (exception) => exception.options.errors,
      ),
      errorCode: buildPlaceholder(
        () => BaseException,
        (exception) => exception.options.errorCode,
      ),
      error: buildPlaceholder(
        () => BaseException,
        (exception) => exception.name,
      ),
      message: buildPlaceholder(
        () => BaseException,
        (exception) => exception.message,
      ),
    },
  },
);
```

But there are exceptions that takes some arguments on building up the error message.

```ts
import { BaseException } from './base.exception';
import { HttpStatus } from '@nestjs/common';
import { ErrorCodes } from 'common/errors/ErrorCodes';
export class ValidationException extends BaseException {
  constructor(errors: any) {
    super('Validation Failed', HttpStatus.BAD_REQUEST, {
      errors,
      errorCode: ErrorCodes.ValidationFailed,
    });
  }
}
```

In the swagger document, they are null or undefined.
How can we set an example with these input params? (like passing a sample error to constructor above)

![image](https://user-images.githubusercontent.com/13379595/133200111-5b917f35-cbb5-4fce-b593-eb6228b76abd.png)

**Comment by kasvith** [comment]
I think I found the solution(its not documented it seems)

```ts
@TemplatedApiException(() => [
    new UserAlreadyRegisteredException('jhondoe@gmail.com'),
  ])
```

We can create a sample errors in usual way, initializing the exception like follows

**Comment by jsproede** [comment]
Hi @kasvith, you're right. Initializing the exception with example values is currently not documented. It's the way, to get example values to the Swagger API documentation. I will add this to the documentation soon :) 

Do you have any further questions?

**Comment by kasvith** [comment]
I found that `externalDocs` in `schema` is not reflecting in the schema

**Comment by jsproede** [comment]
Do you have an example so that I can take a look? We've never used `externalDocs` before :)

**Comment by kasvith** [comment]
For example this one

```ts
@TemplatedApiException(
    () => [new UserAlreadyRegisteredException('jhondoe@gmail.com')],
    {
      schema: {
        description: 'User already registered',
        externalDocs: {
          url: 'https://someplace.com/doc1',
          description: 'Refer this for more',
        },
      },
    },
  )
```

There are two problems I can see

1. The description is not changed even I ve explicitly set it there
![image](https://user-images.githubusercontent.com/13379595/133231233-dd84b857-3d73-4fa5-b966-8e2042f34b99.png)
2. `externalDocs` is missing

```json
       "409": {
            "content": {
              "application/json": {
                "examples": {
                  "UserAlreadyRegisteredException": {
                    "description": "jhondoe@gmail.com already registered",
                    "value": {
                      "message": "jhondoe@gmail.com already registered",
                      "statusCode": 409,
                      "errorCode": 10000,
                      "error": "UserAlreadyRegisteredException"
                    }
                  }
                },
                "schema": {
                  "type": "object",
                  "properties": {
                    "message": {
                      "type": "string",
                      "example": "jhondoe@gmail.com already registered"
                    },
                    "statusCode": {
                      "type": "number",
                      "example": 409
                    },
                    "errorCode": {
                      "type": "number",
                      "example": 10000
                    },
                    "error": {
                      "type": "string",
                      "example": "UserAlreadyRegisteredException"
                    },
                    "errors": {
                      "type": "undefined"
                    }
                  },
                  "required": [
                    "statusCode",
                    "message",
                    "error",
                    "errorCode"
                  ]
                }
              }
            },
            "description": ""
          }
        },
```

**Comment by kasvith** [comment]
I tried this also, still description is not overridden
```ts
@TemplatedApiException(
    () => [new UserAlreadyRegisteredException('jhondoe@gmail.com')],
    {
      description: 'User already registered',
    },
  )
```

**Comment by kasvith** [comment]
Also is there a way to set descriptions for each of these exceptions separately?

```ts
 @TemplatedApiException(() => [
    new UserAlreadyRegisteredException('jhondoe@gmail.com'),
    new UserAlreadyRegisteredException('2'),
  ])
```

By default, it shows the error itself like this
![image](https://user-images.githubusercontent.com/13379595/133233497-0c966124-ce7a-488f-ad36-e4ecb573fde6.png)

**Comment by jsproede** [comment]
Hi @kasvith, thank you for the example. I'll take a look at this issue :)

**Comment by kasvith** [comment]
thank you for this awesome package

**Comment by jsproede** [comment]
Hi @kasvith, you're welcome :) I did take a look at this issue. 
The `schema` property is a bit confusing here, [because it attaches the object to the `message` property](https://nanogiants.github.io/nestjs-swagger-api-exception-decorator/api#schema). But this only works, when using the `@ApiException` decorator with no custom template specified. 

With the latest version of this decorator (`v1.5.0`) it is not possible to enrich the `schema` object with additional properties. 

What do you think about renaming the current `schema` property to `messageSchema` and introduce a new `enrichSchema` property, which allows to enrich/overwrite the generated schema? This would've the following advantages: 
* Although the rename from `schema` to `messageSchema` would be a breaking change, it would still allow to modify the `message` property when no template has been specified
* The new `enrichSchema` property allows to enrich/overwrite the whole auto-generated `schema` with custom attributes

**Comment by kasvith** [comment]
I will think about this and let you know. Maybe I will also try to contribute to this as well :)

**Comment by kasvith** [comment]
Cant we simply merge `schema` properties ?

**Comment by jsproede** [comment]
You mean to use the current `schema` property, and merge its keys and values into the `schema` property in the Swagger documentation? Yes that would be possible, but then anyone who currently uses the existing `schema` property to enrich the `message` property only, is required to change the object passed as `schema` property. Using the `schema` property to specify a custom `message` type, was a refactoring suggestion in issue #1.

**Comment by kasvith** [comment]
I personally don't like overwriting the whole schema there. Most of the time I will want to only change few properties and keep the rest generated by the decorator.

This makes sense because if you want you can do overwriting by writing the entire schema manually there

**Comment by jsproede** [comment]
Both merging and overwriting would be possible with the changed `schema` property :) Only the specified keys would be merged into the `schema` object generated by the decorator. If a key already has been generated by the decorator, with this, it would be possible to overwrite the value :)

**Comment by kasvith** [comment]
Cool, then I think `enrichSchema` would be a good naming choice

**Comment by jsproede** [comment]
Would you still like to contribute? :) Otherwise I would implement the new `enrichSchema` property by the end of this week.

**Comment by kasvith** [comment]
You can go ahead 👍 I won't be able to tackle it this week.

**Comment by jsproede** [comment]
Hi @kasvith, we've just released the version [1.6.0](https://nanogiants.github.io/nestjs-swagger-api-exception-decorator/releasenotes/v1.6.0/). Additionally now the usage of custom exceptions with arguments is documented [here](https://nanogiants.github.io/nestjs-swagger-api-exception-decorator/gettingstarted/usage/custom#custom-exceptions-with-arguments) :)

Thank you for opening this issue. If you've any further questions or ideas to extend this decorator, feel free to open another issue! :)

**Comment by kasvith** [comment]
I just saw it, thank you for spending your time :)
